niceman/resource/singularity.py

```diff
@@ -0,0 +1,240 @@
+"""Environment sub-class to provide management of a Singularity environment."""
+
+import attr
+import os
+import shutil
+import six
+import uuid
+from ..cmd import Runner
+from ..dochelpers import borrowdoc
+from ..support.exceptions import CommandError
+from .session import POSIXSession, Session
+from .base import Resource, attrib
+
+import logging
+lgr = logging.getLogger('niceman.resource.singularity')
+
+
+@attr.s
+class Singularity(Resource):
+    """
+    Environment manager which manages a Singularity container.
+    """
+
+    # Generic properties of any Resource
+    name = attr.ib()
+
+    # Container properties
+    id = attr.ib(default=None)
+    image = attrib(default=None,
+        doc="Singularity image file from which to create the running instance")
+    type = attr.ib(default='singularity')
+
+    status = attr.ib(default=None)
+    _runner = attr.ib(default=None)
+    # A temp image file is needed to get around singularity bug #1185:
+    # https://github.com/singularityware/singularity/issues/1185
+    image_file = attr.ib(default=None)
+
+    def connect(self):
+        """
+        Open a connection to the environment.
+        """
+        self._runner = Runner()
+
+        # Make sure singularity is installed
+        stdout, _ = self._runner.run(['singularity', '--version'])
+        if stdout.startswith('2.2') or stdout.startswith('2.3'):
+            # Running singularity instances and managing them didn't happen
+            # until version 2.4. See: https://singularity.lbl.gov/archive/
+            raise CommandError(msg="Singularity version >= 2.4 required.")
+
+        # Get instance info if we have one running.
+        info = self.get_instance_info()
+        if info:
+            if not self.name:
+                self.name = info['name']
+            if not self.id:
+                # Daemon names for a singularity instance are unique so they
+                # can do double duty as the ID. While not a global ID, it
+                # will do the trick locally.
+                self.id = info['name']
+            self.image = info['image']
+            self.status = 'running'
+        else:
+            self.id = None
+            self.status = None
+
+    def create(self):
+        """
+        Create a container instance.
+
+        Returns
+        -------
+        dict : config parameters to capture in the inventory file
+        """
+        # Check to see if the container is already running.
+        if self.get_instance_info():
+            lgr.info('Resource {} already exists.'.format(self.name))
+        else:
+            self.image_file = "{}.simg".format(str(uuid.uuid4().hex)[:16])
+            if self.image.startswith('shub://') or \
+                    self.image.startswith('docker://'):
+
+                self._runner.run(['singularity', 'pull', '--name',
+                    self.image_file, self.image])
+            else:
+                shutil.copyfile(self.image, self.image_file)
```

Comment:
Hmm, so the code above is to work around the instance.start's "Label string too long" error?  If so, should we avoid copying self.image if its name is short enough?

Reply:
I thought about doing that but opted for keeping the code and process as simple as possible by treating all image names the same. It is assumed that this work around will be removed once the bug is fixed in the singularity code base.